pennylane/gradients/parameter_shift_cv.py

```diff
@@ -0,0 +1,678 @@
+              qml.Beamsplitter(0.5, 0.1, wires=[0, 1])
+              qml.expval(qml.NumberOperator(0))
+
+          tape.trainable_params = {2, 3, 4}
+
+    * If a Fock operation succeeds a Gaussian operation, the Fock operation must
+      not contribute to any measurements. For example, the following is allowed:
+
+      .. code-block:: python
+
+          with qml.tape.JacobianTape() as tape:
+              qml.Displacement(0.6, 0.5, wires=0)
+              qml.Beamsplitter(0.5, 0.1, wires=[0, 1])
+              qml.Kerr(0.654, wires=1)  # there is no measurement on wire 1
+              qml.expval(qml.NumberOperator(0))
+
+          tape.trainable_params = {0, 1, 2}
+
+    If any of the above constraints are not followed, the tape cannot be differentiated
+    via the CV parameter-shift rule. Please use numerical differentiation instead.
+
+    **Example**
+
+    >>> r0, phi0, r1, phi1 = [0.4, -0.3, -0.7, 0.2]
+    >>> dev = qml.device("default.gaussian", wires=1)
+    >>> with qml.tape.JacobianTape() as tape:
+    ...     qml.Squeezing(r0, phi0, wires=[0])
+    ...     qml.Squeezing(r1, phi1, wires=[0])
+    ...     qml.expval(qml.NumberOperator(0))  # second order
+    >>> tape.trainable_params = {0, 2}
+    >>> gradient_tapes, fn = qml.gradients.param_shift_cv(tape, dev)
+    >>> res = dev.batch_execute(gradient_tapes)
+    >>> fn(res)
+    array([[-0.32487113, -0.87049853]])
+    """
+
+    # perform gradient method validation
+    if any(m.return_type is qml.operation.State for m in tape.measurements):
+        raise ValueError(
+            "Computing the gradient of circuits that return the state is not supported."
+        )
+
+    _gradient_analysis(tape)
+
+    gradient_tapes = []
+    shapes = []
+    fns = []
+
+    def _update(data):
+        """Utility function to update the list of gradient tapes,
+        the corresponding number of gradient tapes, and the processing functions"""
+        gradient_tapes.extend(data[0])
+        shapes.append(len(data[0]))
+        fns.append(data[1])
+
+    # TODO: replace the JacobianTape._grad_method_validation
+    # functionality before deprecation.
+    diff_methods = tape._grad_method_validation("analytic" if fallback_fn is None else "best")
+    all_params_grad_method_zero = all(g == "0" for g in diff_methods)
+
+    if not tape.trainable_params or all_params_grad_method_zero:
+        return gradient_tapes, lambda _: np.zeros([tape.output_dim, len(tape.trainable_params)])
+
+    # TODO: replace the JacobianTape._choose_params_with_methods
+    # functionality before deprecation.
+    method_map = dict(tape._choose_params_with_methods(diff_methods, argnum))
+    var_present = any(m.return_type is qml.operation.Variance for m in tape.measurements)
+
+    unsupported_params = [idx for idx, g in method_map.items() if g == "F"]
+    first_order_params = [idx for idx, g in method_map.items() if g == "A"]
+    second_order_params = [idx for idx, g in method_map.items() if g == "A2"]

```

Comment:
(Minor): guess it's not super important, though would having a single for loop be more performant? We seem to be iterating through the same map 3 times.

Reply:
Good point!

Reply:
Have updated this to a single for loop :)